lib/ecto/adapters/mysql/connection.ex

```diff
@@ -11,7 +11,13 @@ if Code.ensure_loaded?(Mariaex.Connection) do
 
     def connect(opts) do
       opts = Keyword.update(opts, :port, @default_port, &normalize_port/1)
-      Mariaex.Connection.start_link(opts)
+      case Mariaex.Connection.start_link(opts) do
+        {:ok, pid} ->
+          repo = Keyword.fetch!(opts, :repo)
+          repo.after_connect(pid)
```

Comment:
We can have a pattern like this:
```elixir
def connect(mod, params, after_connect) do
  case mod.connect(params) do
    {:ok, pid} ->
       after_connect(mod, after_connect, pid)
    {:error, _} = err ->
       err
end

defp after_connect(mod, after_connect, pid) do
   try do
     after_connect.(pid)
   catch
     {class, reason} ->
       mod.disconnect(pid)
       {class, reason, System.stacktrace()}
   else
     _ ->
       :ok   
   end
end
```
Then a pool can call that helper function with the relevant mod/params/fun.

I am slightly tempted to say we should evaluate `after_connect` in a task so we can have a timeout for it and isolate if from the worker so nothing leaks. Lets see what @josevalim thinks.